### PR TITLE
Cdodge/add azure blob storage

### PR DESCRIPTION
**Pull request by chrisndodge**
Required configuration:

changes to lms.auth.json:

New fields:
    "AZURE_ACCOUNT_NAME": <accountName>,
    "AZURE_ACCOUNT_KEY": <accountKey>,
    "AZURE_CONTAINER": <defaultContainer>,

update existing field:
    "DEFAULT_FILE_STORAGE": "storages.backends.azure_storage.AzureStorage", 

changes to lms.env.json:

```
"GRADES_DOWNLOAD": {
    "CONTAINER": <container-to-use-for-instructordashboard>, 
    "STORAGE_TYPE": "azure"
}, 
```

(BTW, the GRADES_DOWNLOAD name is a bit of a legacy misnomer, it is used for all Instructor Dashboard reports)

### CODE REVIEW

**Comment by chenriksson** [comment]
@Microsoft/lex for visibility

**Comment by chrisndodge** [comment]
@chenriksson added configuration information to PR description

File: common/djangoapps/track/management/commands/ship_tracking_logs.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from os.path import isfile, join
+
+from optparse import make_option
+
+from django.core.management.base import BaseCommand
+
+from openedx.core.storage import get_storage
+
+
+class Command(BaseCommand):
```
Comment (chenriksson) [inline common/djangoapps/track/management/commands/ship_tracking_logs.py]: Should we have a help attribute on this class with usage details?

Comment (chenriksson) [inline common/djangoapps/track/management/commands/ship_tracking_logs.py]: Want to confirm usage... would we have nightly cron job on each VM to execute `manage.py ship_tracking_logs`? I'm assuming we'd use folder to differentiate each VM  and avoid filename conflicts?

File: common/djangoapps/track/management/commands/ship_tracking_logs.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                            # remote storage often don't support updates, so we need to
+                            # delete/save pair
+                            storage.delete(dest_fn)
+
+                        print 'Shipping {} to remote storage...'.format(file)
+                        storage.save(dest_fn, f)
+                    else:
+                        print 'File {} already exists in remote storage. Skipping...'.format(file)
+            except Exception, ex:
+                print 'Exception: {}'.format(str(ex))
```
Comment (chenriksson) [inline common/djangoapps/track/management/commands/ship_tracking_logs.py]: Assuming this is routed through rsyslog? Don't suppose you know which service/component (/edx/var/log/*) it falls under?

File: lms/envs/aws.py [hunk 3]
```diff
@@ -428,10 +437,24 @@
 
 AWS_STORAGE_BUCKET_NAME = AUTH_TOKENS.get('AWS_STORAGE_BUCKET_NAME', 'edxuploads')
 
+#
+# To support Azure Storage via the Django-Storages library
+#
+AZURE_ACCOUNT_NAME = AUTH_TOKENS["AZURE_ACCOUNT_NAME"]
+AZURE_ACCOUNT_KEY = AUTH_TOKENS["AZURE_ACCOUNT_KEY"]
+AZURE_CONTAINER = AUTH_TOKENS["AZURE_CONTAINER"]
+
```
Comment (chenriksson) [inline lms/envs/aws.py]: If we want to pull these from server-vars.yml, assume we need to surface them in [config:roles/edxapp/defaults/main.yml](https://github.com/edx/configuration/blob/master/playbooks/roles/edxapp/defaults/main.yml)?

File: lms/envs/aws.py [hunk 4]
```diff
@@ -428,10 +437,24 @@
...
+# normally appends to every returned URL.
+AWS_QUERYSTRING_AUTH = AUTH_TOKENS.get('AWS_QUERYSTRING_AUTH', True)
+AWS_S3_CUSTOM_DOMAIN = AUTH_TOKENS.get('AWS_S3_CUSTOM_DOMAIN', 'edxuploads.s3.amazonaws.com')
+
 if AUTH_TOKENS.get('DEFAULT_FILE_STORAGE'):
     DEFAULT_FILE_STORAGE = AUTH_TOKENS.get('DEFAULT_FILE_STORAGE')
 elif AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY:
     DEFAULT_FILE_STORAGE = 'storages.backends.s3boto.S3BotoStorage'
+elif AZURE_ACCOUNT_NAME and AZURE_ACCOUNT_KEY and AZURE_CONTAINER:
+    DEFAULT_FILE_STORAGE = 'openedx.core.storage.AzureStorageExtended'
```
Comment (chenriksson) [inline lms/envs/aws.py]: I see [PR#11286](https://github.com/edx/edx-platform/pull/11286/files) took a slightly different approach by introducing a new / additive (?) lms/envs for openstack. Is there an advantage to one over the other / should azure eventually have its own envs?

**Comment by chrisndodge** [comment]
Thanks @chenriksson for the feedback. I'll work through the comments.

File: openedx/core/storage.py [hunk 5]
```diff
@@ -30,3 +40,86 @@ class DevelopmentStorage(
...
+    with the same arguments, the same instance is returned. This is useful if
+    the storage implementation makes http requests when instantiated, for
+    example.
+    """
+    return get_storage_class(storage_class)(**kwargs)
+
+
+class AzureStorageExtended(AzureStorage):
+    """
+    A wrapper around the django-stores implementation for Azzure blob storage
```
Comment (chrisndodge) [inline openedx/core/storage.py]: fixed

File: openedx/core/storage.py [hunk 6]
```diff
@@ -30,3 +40,86 @@ class DevelopmentStorage(
...
+    so that it is fully comptaible. The version in the library's repository
```
Comment (chrisndodge) [inline openedx/core/storage.py]: fixed

File: common/djangoapps/track/management/commands/ship_tracking_logs.py (hunk 1 above)
Comment (chrisndodge) [inline common/djangoapps/track/management/commands/ship_tracking_logs.py]: Added help text as suggested

File: common/djangoapps/track/management/commands/ship_tracking_logs.py (hunk 2 above)
Comment (chrisndodge) [inline common/djangoapps/track/management/commands/ship_tracking_logs.py]: Are you referring to the print statements? If so, my experience - but it's not authoritative - when running Django command, this gets written out to stdout. So if you are running this via cron, I'm hoping you can capture this output through cron's normal capture of stdout

File: lms/envs/aws.py (hunk 3 above)
Comment (chrisndodge) [inline lms/envs/aws.py]: Correct. However, a few of these are "secure" configuration (as noted up above as coming from lms.auth.json), so they will have come from your secure private configuration. Be careful not to hard code the Azure secrets in the public "configuration" repo.

Unfortunately, while I've done a lot of plumbing of non-secure configuration in the Ansible Configuration repo, I'm not sure how to do it for secure configuration as the DevOps team at edX did all of that work (developers were not exposed to any production secrets).

File: lms/envs/aws.py (hunk 4 above)
Comment (chrisndodge) [inline lms/envs/aws.py]: Hrm. Let me look a bit more here. Off the cuff I actually don't think this diff here is necessary as I've suggested (up above) that we set DEFAULT_FILE_STORAGE via configuration. This code here - as I understand it - is just to define a default if this is not set at all in lms.auth.json.

File: lms/envs/aws.py (hunk 3 above)
Comment (chenriksson) [inline lms/envs/aws.py]: @chrisndodge so far we've kept automation and (tokenized) config in a public repo, and substitute in secrets from a private repo at deployment time. If you have a better approach, let us know.

Comment (chrisndodge) [inline lms/envs/aws.py]: Thanks. That sounds similar to what edX does. I know they have a private repo for secrets (which I've never seen), however unfortunately I don't know about the mechanics they use to inject those secrets during deployment.

**Pull request merged**